### ISSUE TITLE
torch::save causes serialization error on mnist example

### ISSUE THREAD

**Issue by DrJayLight** [pytorch/pytorch#15777]
## 🐛 Bug

When I add `torch::save(model, "model.pt")` in the mnist example provided here [https://github.com/goldsborough/examples/blob/cpp/cpp/mnist/mnist.cpp](url) it causes a serialization error

## To Reproduce

I modify the example by adding the save functionality as follows:
`for (int32_t epoch = 1; epoch <= options.epochs; ++epoch) {
        train(epoch, options, model, device, *train_loader, optimizer, train_dataset_size);
        test(model, device, *test_loader, test_dataset_size);
        torch::save(model, "model.pt");
    }`

It generates the following error

`In file included from /home/xxxx/experimental_libraries/libtorch/include/torch/csrc/api/include/torch/all.h:8:0,
                 from /home/xxxx/experimental_libraries/libtorch/include/torch/csrc/api/include/torch/torch.h:3,
                 from /home/xxxx/PyTorchProj/Cpp/mnist/mnist.cpp:1:
/home/xxxx/experimental_libraries/libtorch/include/torch/csrc/api/include/torch/serialize.h: In instantiation of ‘void torch::save(const Value&, SaveToArgs&& ...) [with Value = Net; SaveToArgs = {const char (&)[9]}]’:
/home/xxxx/PyTorchProj/Cpp/mnist/mnist.cpp:171:38:   required from here
/home/xxxx/experimental_libraries/libtorch/include/torch/csrc/api/include/torch/serialize.h:39:11: error: no match for ‘operator<<’ (operand types are ‘torch::serialize::OutputArchive’ and ‘const Net’)
   archive << value;
`

## Environment

Python version: 3.6
Is CUDA available: yes
CUDA runtime version: 9.2.148
GPU models and configuration: GPU 0: GeForce GTX 1070
Nvidia driver version: 396.44
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.7.3.1
/usr/local/cuda-9.2/targets/x86_64-linux/lib/libcudnn.so.7.2.1
/usr/local/cuda-9.2/targets/x86_64-linux/lib/libcudnn_static.a

Versions of relevant libraries:
[pip] Could not collect
[conda] blas                      1.0                         mkl  
[conda] magma-cuda90              2.4.0                         1    pytorch
[conda] mkl                       2019.1                      144  
[conda] mkl-include               2019.1                      144  
[conda] mkl-service               1.1.2            py36he904b0f_5  
[conda] mkl_fft                   1.0.6            py36hd81dba3_0  
[conda] mkl_random                1.0.2            py36hd81dba3_0

**Comment by ezyang** [comment]
cc @goldsborough

**Comment by ShigekiKarita** [comment]
You need to wrap with the torch::nn::Module object with torch::nn::ModuleHolder

**Comment by goldsborough** [comment]
The module has to be a `shared_ptr<Net>` instead of just a `Net` in this case. I'll update the tutorial to add checkpointing code.

**Comment by thuyen** [comment]
@goldsborough How about add an overload for `Net` itself, so that we don't have to make a shared_ptr every time.

**Comment by goldsborough** [comment]
@thuyen You're not actually supposed to have to use `shared_ptr`. The C++ frontend tutorial that is about to be published explains this in the section on "Module Ownership": https://github.com/pytorch/tutorials/pull/408

**Comment by thuyen** [comment]
Got it. So we should use module holder even for custom nets then?

**Comment by goldsborough** [comment]
Yes. It's just one extra line. We had to support module holders for internal modules and it became a pain to have overloads for module holders, shared pointers, and regular values for every single API method. So module holders are the recommended API

**Comment by CDitzel** [comment]
I always despise the use of macros to facilitate certain aspects that should be language inherent but are limited by the language at hand. One popular example is the infamous Q_OBJECT macro of QT to enable the signal slot mechanics. 

So there are situations in which macros seem to be the only way to accomplish a certain goals, although it causes some boiler plate code. I do however appreciate that the C++ front end is written in a very concise and modern C++ which makes it rather easy to follow and understand crucial aspects of the code base
